apis/v1alpha1/httproute_types.go

```diff
@@ -165,32 +180,68 @@ type HTTPRouteMatch struct {
 	ExtensionRef *RouteMatchExtensionObjectReference `json:"extensionRef" protobuf:"bytes,5,opt,name=extensionRef"`
 }
 
-// HTTPRouteFilter defines a filter-like action to be applied to
-// requests.
-type HTTPRouteFilter struct {
-	// Headers related filters.
-	//
-	// Support: extended
+const (
+	// FilterHTTPReqeustHeaderTransformer can be used to add or remove an HTTP
+	// header from an HTTP request before it is sent to the upstream target.
+	// Support: core
 	// +optional
-	Headers *HTTPHeaderFilter `json:"headers" protobuf:"bytes,1,opt,name=headers"`
+	FilterHTTPReqeustHeaderTransformer = "HTTPRequestHeaderTransformer"
+	// TODO(hbagdi): add FilterHTTPRequestMirroring filter
+)
+
+// HTTPRouteFilter defines additional processing steps that must be completed
+// during the request or response lifecycle.
+// HTTPRouteFilters are meant as an extension point to express additional
+// processing that must be done in Gateway implementations. Some examples include
+// request or response modification, implementing authentication strategies,
+// rate-limiting, and traffic shaping.
+// API guarantee/conformance is defined based on the type of the filter.
+// +union
+type HTTPRouteFilter struct {
+	// Type identifies the filter to execute.
+	// Types are classified into two conformance-levels:
+	// - Core and extended: These filter types and their corresponding configuration
+	//   is defined in this package. All implementations must implement
+	//   the core filters. Implementers are encouraged to support extended filters.
+	//   Definitions for filter-specific configuration for these
+	//   filters is defined in this package.
+	// - Custom: These filters are defined and supported by specific vendors.
```

Comment:
Does this mean for each filter implemented by a vendor, a corresponding type would be added into the above const list? Or we can have just one single type, for example "ImplementationSpecific", and the actual type of the filter can be figured out from the kind of the CRD resource?

Reply:
I had intentionally kept the `Type` field in HTTPRouteFilter as `string` to allow for free-form values for custom filters and the constants declared in this file could be used for core filter types. This was going to be useful with `Config` type, which based on feedback, we are going to get rid off. 

Now considering these changes, and looking at other places where we follow this pattern, it makes sense to have a single "ImplementationSpecific" filter type for all custom filters.
WDYT?

Reply:
+1 on having a single type

Reply:
Updated!